---
title: "CIT analysis"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/microglia_analysis
ml git proj gdal geos pandoc
git pull
R

system("git pull")


rmarkdown::render("EPHA1_AS1.Rmd")


--->

```{r load.packages}
suppressPackageStartupMessages({
library(data.table)
library(ggplot2)
library(ggrepel)
library(org.Hs.eg.db)
library(GenomicFeatures)
library(dplyr)
library(stringr) 
library(knitr)
library(kableExtra)
library(decorate)
library(MeSH.db)
library(synapser)
library(data.table)
library(GenomicFeatures)
library(ggplot2)
})

synLogin()

src = '/hpc/users/hoffmg01/build2/mmQTL'

source(paste0(src, "/make_plots.R"))
source(paste0(src, "/plot_genes.R"))

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c('png', 'pdf'),
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r caQTL.data}
# Read SNP locations: hg19
df_snp = fread( synGet('syn24912590')$path )
colnames(df_snp) = c("Chr", "Position0", "Position", 'ID')
df_snp[,Position0 := c()]
df_snp[,Chr:=gsub("^chr", '', Chr)]
setkey(df_snp, 'ID')
# gc()

# Read xQTL p-values
df_caQTL = fread( synGet('syn25000646')$path, header=TRUE )
df_caQTL[,Chr:=gsub("^chr", '', Chr)]
colnames(df_caQTL)[colnames(df_caQTL) == 'caQTL_order'] = 'eQTL_order'
df_caQTL[,log10.p.value := -1*pnorm(abs(Z_score_fixed), 0, 1, lower.tail=FALSE, log.p=TRUE) / log(10) - log10(2)]
df_caQTL[,Z_score_fixed:=c()]
df_caQTL[,Z_score_random:=c()]
colnames(df_caQTL)[colnames(df_caQTL)=="Peak"] = "Gene"
setkey(df_caQTL, "Variant")
gc()

# Include Position in df_caQTL
df_caQTL = merge(df_caQTL, df_snp[,c("Position", "ID")], by.x="Variant", by.y="ID")
setkey(df_caQTL, Position)
gc()

# Read fine-mapping results
df_finemap = fread( synGet('syn24964518')$path)
# colnames(df_finemap) = c("Chr", "Gene", "eQTL_order", "Variant", "PIP")
colnames(df_finemap)[colnames(df_finemap)=="PP"] = "PIP"
df_finemap[,Chr := c()]
setkey(df_finemap, 'Variant')

# Merg SNP locations and fine-mapping
df_finemap = merge(df_finemap, df_snp, by.x="Variant", by.y="ID")
setkey(df_finemap, Position)

df_caQTL = merge(df_caQTL, df_finemap[,c("Variant", "PIP", "Gene")], by=c("Variant", "Gene"), all.x=TRUE)

# get peak locations: hg19
df_peak_location = fread(synGet('syn25944661')$path)
colnames(df_peak_location) = c("Chr", "start", "end", "name")
gr_peak_location = with(df_peak_location, GRanges(Chr, IRanges(start, end, name=name)))
```


```{r eQTL.data}
# Read SNP locations
df_snp = fread( synGet('syn24196924')$path )
colnames(df_snp) = c("Chr", "Position0", "Position", 'ID')
df_snp[,Position0 := c()]
df_snp[,Chr:=gsub("^chr", '', Chr)]
setkey(df_snp, 'ID')
gc()

# Read xQTL p-values
df_eqtl = fread( synGet('syn24182390')$path, header=FALSE )
df_eqtl[,V7:=c(),]
cn = fread(synGet('syn24182390')$path, nrow=0)
colnames(df_eqtl) = colnames(cn)
df_eqtl[,Chr:=gsub("^chr", '', Chr)]

df_eqtl[,log10.p.value := -1*pnorm(Z_score_random, 0, 1, lower.tail=FALSE, log.p=TRUE) / log(10) - log10(2)]
df_eqtl[,Z_score_fixed:=c()]
df_eqtl[,Z_score_random:=c()]
setkey(df_eqtl, "Variant")
gc()

# Include Position in df_eqtl
df_eqtl = merge(df_eqtl, df_snp[,c("Position", "ID")], by.x="Variant", by.y="ID")
setkey(df_eqtl, Position)
gc()

# Read fine-mapping results
df_finemap = fread( synGet('syn24178479')$path )
colnames(df_finemap)[5] = "PIP"
df_finemap[,Chr := c()]
setkey(df_finemap, 'Variant')

# Merg SNPe locations and fine-mapping
df_finemap = merge(df_finemap, df_snp, by.x="Variant", by.y="ID")
setkey(df_finemap, Position)

df_eqtl = merge(df_eqtl, df_finemap[,c("Variant", "PIP", "Gene")], by=c("Variant", "Gene"), all.x=TRUE)
```			


```{r causalDB}
# read CAUSALdb 
###############
# /Users/gabrielhoffman/workspace/scripts/compositeQTL/causaDB.R

# This combines the results from CausalDB: http://mulinlab.tmu.edu.cn/causaldb/index.html
causalDB = readRDS( synGet('syn24178554')$path )

# set keys for faster searching
setkey(causalDB[["credible_set"]], "meta_id")
setkey(causalDB[['meta_info']], "ID")

causalDB[['meta_info']]["OT344",]

# Find which study to plot
df_causaldb = data.table(causalDB[["credible_set"]])
df_causaldb[,Variant:=paste0('rs', rsID)]
df_ad_finemap = df_causaldb[meta_id == 'OT344',]
```

```{r read.microglia.finemapping}
df_fm_caQTL = fread("/hpc/users/hoffmg01/work/mmQTL/microglia/caqtl/figures/finemap_caQTL.csv")
df_fm_eQTL = fread("/hpc/users/hoffmg01/work/mmQTL/microglia/eqtl/figures/finemap_eQTL.csv")

df_fm_caQTL = df_fm_caQTL[PMID %in% c(30617256),] # 29777097
df_fm_eQTL = df_fm_eQTL[PMID %in% c(30617256),]# 29777097

df_fm_caQTL = merge(df_fm_caQTL, unique(df_caQTL[,c("Variant", "Chr", "Position")]),by="Variant")
df_fm_eQTL = merge(df_fm_eQTL, unique(df_eqtl[,c("Variant", "Chr", "Position")]),by="Variant")
```


```{r load.data}
# Load AD GWAS
AD_GWAS_summary = fread("/sc/arion/projects/roussp01a/mads_atacseq/data/gwas/alz2noapoe/AD_sumstats_Jansenetal_NO_APO_E.txt.gz", header=FALSE)
colnames(AD_GWAS_summary) = c('uniqID.a1a2', 'CHR', 'BP', 'A1', 'A2', 'SNP', 'Z', 'P', 'Nsum', 'Neff' , 'dir', 'MAF', 'BETA', 'SE')
```

# save(list=ls(), file="AD_all.RDATA")
# load("AD_all.RDATA")

```{r make.plots}
plot_triple = function(SNP, ensGene, peakID,window, fileOut){

	chrom = as.numeric(unique(df_eqtl[Variant == SNP,'Chr']))
	center = as.numeric(unique(df_eqtl[Variant == SNP,'Position']))

	wh = GRanges(paste0("chr", chrom), IRanges(center-window, center+window))

	geneInfo = select(EnsDb.Hsapiens.v75, keys=ensGene, keytype="GENEID", column=c('GENENAME', 'GENEBIOTYPE'))
	Symbol = geneInfo$GENENAME

	edges = c(start(wh), end(wh))

	# AD GWAS
	df = AD_GWAS_summary[CHR == chrom,]
	df = df[(BP >= center-window) & (BP <= center+window),]

	df2 = merge(df, df_ad_finemap[,c('Variant', 'FINEMAP')], by.x="SNP", by.y="Variant", all.x=TRUE)
	df2[,isCandSet :=!is.na(FINEMAP)]

	ymax = -log10(min(df$P))

	fig_AD_GWAS = ggplot(df2, aes(BP, -log10(P), color=!is.na(FINEMAP), size=isCandSet+1.0 )) + geom_point() + scale_x_continuous(limits=edges, expand=c(0,0)) + scale_y_continuous(expand=c(0, 0), limits=c(0,ymax*1.03)) + scale_color_manual(values=c("black", "red")) + ylab(bquote(-log[10]~P)) + scale_size(range=c(.8, 2), breaks=c(1.1))


	# eQTL
	######
	df = df_eqtl[Chr == chrom,]
	df = df[(Position >= center-window) & (Position <= center+window),]
	df = df[Gene==ensGene,]
	df[,isCandSet:=!is.na(PIP)]

	sizeRng = c(.8)
	if( any(df$isCandSet) ){
		sizeRng = append(sizeRng, 2)
	}

	# Get CLPP from eQTL
	clpp_e = df_fm_eQTL[(Variant==SNP)&(Gene==ensGene),data.table(Author=`Consortium/author`, CLPP=PIP.prod)]
	clpp_e = paste(with(clpp_e, paste0('CLPP=', format(CLPP, digits=2), ' ')), collapse="\n")

	if(clpp_e == "CLPP= "){
		clpp_e = ''
	}

	ymax = max(df$log10.p.value)
	fig_eQTL = ggplot(df, aes(Position, log10.p.value, color=!is.na(PIP), size=isCandSet+1.0)) + geom_point() + scale_x_continuous(limits=edges, expand=c(0,0)) + scale_y_continuous(expand=c(0, 0), limits=c(0,ymax*1.03)) + scale_color_manual(values=c("black", "red")) + ylab(bquote(-log[10]~P)) + scale_size(range=sizeRng, breaks=c(1.1)) + annotate("text",  x=edges[2], y = ymax, label = clpp_e, vjust=1, hjust=1)

	# caQTL
	#######
	df = df_caQTL[Chr == chrom,]
	df = df[(Position >= center-window) & (Position <= center+window),]
	df = df[Gene==peakID,]

	if( nrow(df) > 0){
		df[,isCandSet:=!is.na(PIP)]

		sizeRng = c(.8)
		if( any(df$isCandSet) ){
			sizeRng = append(sizeRng, 2)
		}

		# Get CLPP from caQTL
		clpp_ca = df_fm_caQTL[(Variant==SNP)&(Gene==peakID),data.table(Author=`Consortium/author`, CLPP=PIP.prod)]
		clpp_ca = paste(with(clpp_ca, paste0('CLPP=', format(CLPP, digits=2), ' ')), collapse="\n")

		if(clpp_ca == "CLPP= "){
			clpp_ca = ''
		}

		ymax = max(df$log10.p.value)
		fig_caQTL = ggplot(df, aes(Position, log10.p.value, color=!is.na(PIP), size=isCandSet+1)) + geom_point() + scale_x_continuous(limits=edges, expand=c(0,0)) + scale_y_continuous(expand=c(0, 0), limits=c(0,ymax*1.03)) + scale_color_manual(values=c("black", "red")) + ylab(bquote(-log[10]~P)) + scale_size(range=sizeRng, breaks=c(1.1)) + annotate("text",  x=edges[2], y = ymax, label = clpp_ca, vjust=1, hjust=1)
		}else{
			fig_caQTL = ggplot()
		}
	# pkgload::unload("EnsDb.Hsapiens.v75")
	# library(EnsDb.Hsapiens.v75)

	fig_genebody = plotEnsGenes_gg( EnsDb.Hsapiens.v75, start(wh), end(wh), seqnames(wh), splice_variants=FALSE, non_coding=TRUE)

	# ATAC
	######

	gr_loc = gr_peak_location[findOverlaps(wh, gr_peak_location)@to]

	fig_ATAC = ggplot() + scale_x_continuous(expand=c(0,0), limits=edges) + scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_color_manual(values = c("black", "red")) + theme_bw() + theme(legend.position = "none", axis.title=element_blank(), axis.ticks.y=element_blank(), axis.line.y=element_blank(), panel.grid=element_blank(), axis.text.y=element_blank(),panel.background = element_blank(), strip.background = element_blank(), rect = element_rect(fill="white", linetype=0))
	if( length(gr_loc) > 0){
		df_loc = as.data.frame(gr_loc)

		col = c("black", "red")
		df_loc$status = 0
		df_loc$status[rownames(df_loc) == peakID] = 1
		df_loc$status = factor(as.character(df_loc$status), c(0, 1))

		col = c()
		if( any(df_loc$status==0)){
			col = append(col, "black")
		}
		if( any(df_loc$status==1)){
			col = append(col, "red")
		}

		df_loc$start = pmax(df_loc$start, edges[1])
		df_loc$end = pmin(df_loc$end, edges[2])


		fig_ATAC = ggplot(df_loc) + geom_segment(aes(x=start, y=1, xend = end, yend=1, color=status), size=4) + scale_x_continuous(expand=c(0,0), limits=edges) + scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_color_manual(values = col) + theme_bw() + theme(legend.position = "none", axis.title=element_blank(), axis.ticks.y=element_blank(), axis.line.y=element_blank(), panel.grid=element_blank(), axis.text.y=element_blank(),panel.background = element_blank(), strip.background = element_blank(), rect = element_rect(fill="white", linetype=0))
	}

	fig_track = ggbio::tracks( "AD GWAS" = fig_AD_GWAS, 
											"eQTL" 		= fig_eQTL,
											"caQTL" 	= fig_caQTL,
											"Genes"   = fig_genebody,
											"OCR" 		= fig_ATAC,
											xlim = wh,
											padding = unit(-.65, "lines"),
											label.bg.fill="navy", label.text.color="white",
											heights=c(1, 1, 1, .4, .3),
											theme = theme_bw(8) + theme(legend.position="none", panel.grid.minor = element_blank(), panel.grid.major = element_blank()),
							title=Symbol )
	fig_track@mutable['Genes'] = FALSE
	fig_track@mutable['OCR'] = FALSE
	ggbio::ggsave(fig_track, file=fileOut)
}
```





```{r make_plots}
# Plot Locations
# EPHA1-AS1

SNP = 'rs11771145'
ensGene = 'ENSG00000229153'
peakID = "Peak_188003"
plot_triple(SNP, ensGene, peakID, window = 80000, "figures/EPHA1-AS1.pdf")


# PICALM
SNP = 'rs10792832'
ensGene = 'ENSG00000073921'
peakID = "Peak_30728"
plot_triple(SNP, ensGene, peakID, window = 150000, "figures/PICALM.pdf")

# For each eQTL
for(i in 1:nrow(df_fm_eQTL)){

		SNP = df_fm_eQTL[i,Variant]
		ensGene = df_fm_eQTL[i,Gene]
		Symbol = df_fm_eQTL[i,Symbol]

		# get peak where this SNP has the lowest p-value
		peakID = df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",Symbol, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


# For each caQTL
for(i in 1:nrow(df_fm_caQTL)){

		SNP = df_fm_caQTL[i,Variant]
		peakID = df_fm_caQTL[i,Gene]

		# get peak where this SNP has the lowest p-value
		ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",peakID, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


  
file = "/sc/arion/projects/Microglia/Combined_RNAseq_ATACseq/GWAS_loci_investigations/GWAS_loci_analyses/Novel_Loci/NovelLoci_locations_RK_6_24_21.csv"
df_focus = fread(file)

# For each eQTL
df_focus_gene = df_focus[Gene_ID !='',]

for(i in 1:nrow(df_focus_gene)){

	SNP = df_focus_gene[i,SNP_ID]
	ensGene = df_focus_gene[i,Gene_ID]
	Symbol = df_focus_gene[i,GeneName]

	# get peak where this SNP has the lowest p-value
	peakID =	df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

	if( length(peakID) == 0){
		peakID = ''
	}
	file = paste0("figures/",Symbol, "_", SNP, "_fm.pdf")
	plot_triple(SNP, ensGene, peakID, window = 80000, file)
}



# For each caQTL
df_focus_peak = df_focus[Peak_ID !='',]

for(i in 1:nrow(df_focus_peak)){

		SNP = df_focus_peak[i,SNP_ID]
		peakID = df_focus_peak[i,Peak_ID]

		# get peak where this SNP has the lowest p-value
		ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",peakID, "_", SNP, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}

```


