---
title: "CIT analysis"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/microglia_analysis
ml git proj gdal geos pandoc
git pull
R

system("git pull")


rmarkdown::render("EPHA1_AS1.Rmd")




--->

```{r load.packages}
suppressPackageStartupMessages({
library(data.table)
library(ggplot2)
library(ggrepel)
library(org.Hs.eg.db)
library(GenomicFeatures)
library(dplyr)
library(stringr) 
library(knitr)
library(kableExtra)
# library(decorate)
library(MeSH.db)
library(synapser)
library(data.table)
library(GenomicFeatures)
library(ggplot2)
})

synLogin()

src = '/hpc/users/hoffmg01/build2/mmQTL'

source(paste0(src, "/make_plots.R"))
source(paste0(src, "/plot_genes.R"))
source("plot_triple.R")

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c('png', 'pdf'),
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r caQTL.data}
# Read SNP locations: hg19
df_snp = fread( synGet('syn24912590')$path )
colnames(df_snp) = c("Chr", "Position0", "Position", 'ID')
df_snp[,Position0 := c()]
df_snp[,Chr:=gsub("^chr", '', Chr)]
setkey(df_snp, 'ID')
# gc()

# Read xQTL p-values
df_caQTL = fread( synGet('syn25000646')$path, header=TRUE )
df_caQTL[,Chr:=gsub("^chr", '', Chr)]
colnames(df_caQTL)[colnames(df_caQTL) == 'caQTL_order'] = 'eQTL_order'
df_caQTL[,log10.p.value := -1*pnorm(abs(Z_score_fixed), 0, 1, lower.tail=FALSE, log.p=TRUE) / log(10) - log10(2)]
df_caQTL[,Z_score_fixed:=c()]
df_caQTL[,Z_score_random:=c()]
colnames(df_caQTL)[colnames(df_caQTL)=="Peak"] = "Gene"
setkey(df_caQTL, "Variant")
gc()

# Include Position in df_caQTL
df_caQTL = merge(df_caQTL, df_snp[,c("Position", "ID")], by.x="Variant", by.y="ID")
setkey(df_caQTL, Position)
gc()

# Read fine-mapping results
df_finemap = fread( synGet('syn24964518')$path)
# colnames(df_finemap) = c("Chr", "Gene", "eQTL_order", "Variant", "PIP")
colnames(df_finemap)[colnames(df_finemap)=="PP"] = "PIP"
df_finemap[,Chr := c()]
setkey(df_finemap, 'Variant')

# Merg SNP locations and fine-mapping
df_finemap = merge(df_finemap, df_snp, by.x="Variant", by.y="ID")
setkey(df_finemap, Position)

df_caQTL = merge(df_caQTL, df_finemap[,c("Variant", "PIP", "Gene")], by=c("Variant", "Gene"), all.x=TRUE)

# get peak locations: hg19
df_peak_location = fread(synGet('syn25944661')$path)
colnames(df_peak_location) = c("Chr", "start", "end", "name")
gr_peak_location = with(df_peak_location, GRanges(Chr, IRanges(start, end, name=name)))
```


```{r eQTL.data}
# Read SNP locations
df_snp = fread( synGet('syn24196924')$path )
colnames(df_snp) = c("Chr", "Position0", "Position", 'ID')
df_snp[,Position0 := c()]
df_snp[,Chr:=gsub("^chr", '', Chr)]
setkey(df_snp, 'ID')
# gc()

# Read xQTL p-values
df_eqtl = fread( synGet('syn24182390')$path, header=FALSE )
df_eqtl[,V7:=c(),]
cn = fread(synGet('syn24182390')$path, nrow=0)
colnames(df_eqtl) = colnames(cn)
df_eqtl[,Chr:=gsub("^chr", '', Chr)]

df_eqtl[,log10.p.value := -1*pnorm(Z_score_random, 0, 1, lower.tail=FALSE, log.p=TRUE) / log(10) - log10(2)]
df_eqtl[,Z_score_fixed:=c()]
df_eqtl[,Z_score_random:=c()]
setkey(df_eqtl, "Variant")
# gc()

# Include Position in df_eqtl
df_eqtl = merge(df_eqtl, df_snp[,c("Position", "ID")], by.x="Variant", by.y="ID")
setkey(df_eqtl, Position)
# gc()

# Read fine-mapping results
df_finemap = fread( synGet('syn24178479')$path )
colnames(df_finemap)[5] = "PIP"
df_finemap[,Chr := c()]
setkey(df_finemap, 'Variant')

# Merg SNPe locations and fine-mapping
df_finemap = merge(df_finemap, df_snp, by.x="Variant", by.y="ID")
setkey(df_finemap, Position)

df_eqtl = merge(df_eqtl, df_finemap[,c("Variant", "PIP", "Gene")], by=c("Variant", "Gene"), all.x=TRUE)
```			


```{r causalDB}
# read CAUSALdb 
###############
# /Users/gabrielhoffman/workspace/scripts/compositeQTL/causaDB.R

# This combines the results from CausalDB: http://mulinlab.tmu.edu.cn/causaldb/index.html
causalDB = readRDS( synGet('syn24178554')$path )

# set keys for faster searching
setkey(causalDB[["credible_set"]], "meta_id")
setkey(causalDB[['meta_info']], "ID")

causalDB[['meta_info']]["OT344",]

# Find which study to plot
df_causaldb = data.table(causalDB[["credible_set"]])
df_causaldb[,Variant:=paste0('rs', rsID)]
df_ad_finemap = df_causaldb[meta_id == 'OT344',]
```

```{r read.microglia.finemapping}
df_fm_caQTL = fread("/hpc/users/hoffmg01/work/mmQTL/microglia/caqtl/figures/finemap_caQTL.csv")
df_fm_eQTL = fread("/hpc/users/hoffmg01/work/mmQTL/microglia/eqtl/figures/finemap_eQTL.csv")

df_fm_caQTL = df_fm_caQTL[PMID %in% c(30617256),] # 29777097
df_fm_eQTL = df_fm_eQTL[PMID %in% c(30617256),]# 29777097

df_fm_caQTL = merge(df_fm_caQTL, unique(df_caQTL[,c("Variant", "Chr", "Position")]),by="Variant")
df_fm_eQTL = merge(df_fm_eQTL, unique(df_eqtl[,c("Variant", "Chr", "Position")]),by="Variant")
```


```{r load.data}
# Load AD GWAS
AD_GWAS_summary = fread("/sc/arion/projects/roussp01a/mads_atacseq/data/gwas/alz2noapoe/AD_sumstats_Jansenetal_NO_APO_E.txt.gz", header=FALSE)
colnames(AD_GWAS_summary) = c('uniqID.a1a2', 'CHR', 'BP', 'A1', 'A2', 'SNP', 'Z', 'P', 'Nsum', 'Neff' , 'dir', 'MAF', 'BETA', 'SE')
```

# save(list=ls(), file="AD_all.RDATA")
# load("AD_all.RDATA")


```{r exit, cache=FALSE}
knitr::knit_exit()
```

```{r qtlPlots}

library(qtlPlots)

# GWAS
df_merge = merge(AD_GWAS_summary, df_ad_finemap[,c('Variant', 'FINEMAP')], by.x="SNP", by.y="Variant", all.x=TRUE)
df_merge[,inCandidateSet :=!is.na(FINEMAP)]

gr_ad_gwas = with(df_merge, GRanges(paste0('chr',CHR), IRanges(start=BP, end=BP, score=-log10(P), PIP = FINEMAP, inCandidateSet=inCandidateSet)))

# eQTL
gr_eqtl = with(df_eqtl[eQTL_order==1,], GRanges( Chr, IRanges(start=Position, end=Position, score=log10.p.value, Variant = Variant, Gene=Gene, PIP=PIP, inCandidateSet=!is.na(PIP)) ))

# df_caQTL
gr_caqtl = with(df_caQTL[eQTL_order==1,], GRanges( Chr, IRanges(start=Position, end=Position, score=log10.p.value, Variant = Variant, Gene=Gene, PIP=PIP,, inCandidateSet=!is.na(PIP)) ))

# ATAC-seq peaks
gr_peaks = with(df_peak_location, GRanges(Chr, IRanges(start, end, name=name)))

devtools::unload("EnsDb.Hsapiens.v75")

ensdb = EnsDb.Hsapiens.v75::EnsDb.Hsapiens.v75 # genes on hg19


plot_combine = function(wh, ensGene, peakID, SNP){

	# GWAS
	fig_gwas = plotMht( gr_ad_gwas, wh )

	# eQTL 
	fig_eqtl = plotMht( gr_eqtl[gr_eqtl$Gene == ensGene], wh )

	# caQTL 
	fig_caqtl = plotMht( gr_caqtl[gr_caqtl$Gene == peakID], wh )


	df = merge( data.frame(gr_ad_gwas), data.frame(gr_eqtl[gr_eqtl$Gene == ensGene]), by="start")
	df$CLPP.eqtl = with(df, PIP.x*PIP.y)
	df = df[!is.na(df$CLPP.eqtl),]
	df_eqtl = data.frame(Variant = df$Variant, 'Gene expression' = df$CLPP.eqtl, check.names=FALSE)

	df = merge( data.frame(gr_ad_gwas), data.frame(gr_caqtl[gr_caqtl$Gene == peakID]), by="start")
	df$CLPP.caqtl = with(df, PIP.x*PIP.y)
	df = df[!is.na(df$CLPP.caqtl),]
	df_caqtl = data.frame(Variant = df$Variant, 'Chromatin accessability' = df$CLPP.caqtl, check.names=FALSE)

	df_clpp = merge( df_eqtl, df_caqtl, by="Variant")

	df = reshape2::melt(df_clpp, id.vars="Variant")
	fig_CLPP = ggplot(df, aes(Variant, value, fill=variable)) + geom_bar(stat="identity") + theme_classic() + scale_fill_brewer(name="Assay", palette="Set1") + ylab("CLPP") + scale_y_continuous(expand=c(0, 0), limits=c(0, NA))


	# Gene
	fig_gene = plotEnsGenes( ensdb, wh)

	# ATAC-seq peaks
	# highlight the target peak
	gr_peaks$status = 0
	gr_peaks$status[names(gr_peaks) == peakID] = 1
	fig_bed = plotBed( gr_peaks, wh )

	# CAUSALdb plot
	# blid = df_causaldb[Variant == SNP,unique(block_id)]
	# df = df_causaldb[block_id == blid,]
	# df = merge(df, causalDB[['meta_info']], by.x="meta_id", by.y="ID")
	# df = df[grep("Alz", df$MeSH_term),]
	# df = df[!df$meta_id %in% c('CA184'),]

	# fig_cdb = ggplot(df, aes(BP, FINEMAP, color = paste0(Trait, ' (',`Consortium/author`, ', ', Year,')'))) + geom_point() + scale_x_continuous(expand=c(0,0), limits=c(start(wh), end(wh))) + theme_classic() + scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_color_brewer(name="Study", palette="Set1") + theme(legend.position="none")

	fig_track = ggbio::tracks(
			"AD GWAS"				= fig_gwas,
	     "eQTL"        = fig_eqtl, 
	    "caQTL"				= fig_caqtl,
	    # "causalDB"   = fig_cdb,
	    'Genes'          = fig_gene,
	    'OCR'          = fig_bed,
	    xlim = wh,
	    padding = unit(-.65, "lines"), 
	    label.bg.fill="navy", 
	    label.text.color="white",
	    heights=c(1, 1, 1,.25, .1),
	    title = peakID) 
	fig_track@mutable['Genes'] = FALSE
	
	list(fig_track = fig_track, fig_CLPP = fig_CLPP)
}



# PICALM
SNP = 'rs10792832'
ensGene = 'ENSG00000073921'
peakID = "Peak_30728"
window = 180000

gr = gr_eqtl[gr_eqtl$Variant == SNP]

chrom = as.character(unique(seqnames(gr)@values))
center = unique(start(gr))
wh = GRanges(paste0("chr", chrom), IRanges(center-window, center+window))

figList = plot_combine(wh, ensGene, peakID, SNP)

fileDir = "/hpc/users/hoffmg01/work/microglia_analysis/figures/qtlPlots"

file = paste0(fileDir, '/', ensGene, "_", SNP, ".pdf")
pdf(file)
figList$fig_track
figList$fig_CLPP
dev.off()




# 
# file = "/sc/arion/projects/Microglia/Combined_RNAseq_ATACseq/GWAS_loci_investigations/GWAS_loci_analyses/Novel_Loci/AD_Loci_locations_RK_7_8_21.csv"  
file = "/sc/arion/projects/Microglia/Combined_RNAseq_ATACseq/GWAS_loci_investigations/GWAS_loci_analyses/Novel_Loci/NovelLoci_locations_RK_6_24_21.csv"
df_focus = fread(file)

fileDir = "/hpc/users/hoffmg01/work/microglia_analysis/figures/qtlPlots"
window = 180000

# For each eQTL
df_focus_gene = df_focus[Gene_ID !='',]

for(i in 1:nrow(df_focus_gene)){

	SNP = df_focus_gene[i,SNP_ID]
	ensGene = df_focus_gene[i,Gene_ID]
	Symbol = df_focus_gene[i,GeneName]

	# get peak where this SNP has the lowest p-value
	peakID =df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

	if( length(peakID) == 0){
		peakID = ''
	}

	gr = gr_eqtl[gr_eqtl$Variant == SNP]

	chrom = as.character(unique(seqnames(gr)@values))
	center = unique(start(gr))
	wh = GRanges(paste0("chr", chrom), IRanges(center-window, center+window))

	figList = plot_combine(wh, ensGene, peakID, SNP)

	file = paste0(fileDir, '/', Symbol, "_", ensGene, "_", peakID, "_",SNP, ".pdf")
	pdf(file)
	figList$fig_track
	figList$fig_CLPP
	dev.off()
}


# For each caQTL
window = 80000
df_focus_peak = df_focus[Peak_ID !='',]

for(i in 1:nrow(df_focus_peak)){

	SNP = df_focus_peak[i,SNP_ID]
	peakID = df_focus_peak[i,Peak_ID]

	# get peak where this SNP has the lowest p-value
	ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

	gr = gr_eqtl[gr_eqtl$Variant == SNP]

	chrom = as.character(unique(seqnames(gr)@values))
	center = unique(start(gr))
	wh = GRanges(paste0("chr", chrom), IRanges(center-window, center+window))

	figList = plot_combine(wh, ensGene, peakID, SNP)

	file = paste0(fileDir, '/', peakID, "_", SNP, ".pdf")
	pdf(file)
	figList$fig_track
	figList$fig_CLPP
	dev.off()

}

```




```{r make_plots, cache=FALSE}

pkgload::unload("EnsDb.Hsapiens.v75")
library(EnsDb.Hsapiens.v75)

source("plot_triple.R")

# Plot Locations
# EPHA1-AS1
SNP = 'rs11771145'
ensGene = 'ENSG00000229153'
peakID = "Peak_188003"
plot_triple(SNP, ensGene, peakID, window = 50000, "figures/EPHA1-AS1.pdf")




# system("git pull"); source("plot_triple.R")
# plot_triple(SNP, ensGene, peakID, window = 50000, "figures/EPHA1-AS1.pdf")




# PICALM
SNP = 'rs10792832'
ensGene = 'ENSG00000073921'
peakID = "Peak_30728"
plot_triple(SNP, ensGene, peakID, window = 150000, "figures/PICALM.pdf")













# For each eQTL
for(i in 1:nrow(df_fm_eQTL)){

		SNP = df_fm_eQTL[i,Variant]
		ensGene = df_fm_eQTL[i,Gene]
		Symbol = df_fm_eQTL[i,Symbol]

		# get peak where this SNP has the lowest p-value
		peakID = df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",Symbol, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


# For each caQTL
for(i in 1:nrow(df_fm_caQTL)){

		SNP = df_fm_caQTL[i,Variant]
		peakID = df_fm_caQTL[i,Gene]

		# get peak where this SNP has the lowest p-value
		ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",peakID, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


  
file = "/sc/arion/projects/Microglia/Combined_RNAseq_ATACseq/GWAS_loci_investigations/GWAS_loci_analyses/Novel_Loci/NovelLoci_locations_RK_6_24_21.csv"
df_focus = fread(file)

# For each eQTL
df_focus_gene = df_focus[Gene_ID !='',]

for(i in 1:nrow(df_focus_gene)){

	SNP = df_focus_gene[i,SNP_ID]
	ensGene = df_focus_gene[i,Gene_ID]
	Symbol = df_focus_gene[i,GeneName]

	# get peak where this SNP has the lowest p-value
	peakID =	df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

	if( length(peakID) == 0){
		peakID = ''
	}
	file = paste0("figures/",Symbol, "_", SNP, "_fm.pdf")
	plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


# For each caQTL
df_focus_peak = df_focus[Peak_ID !='',]

for(i in 1:nrow(df_focus_peak)){

		SNP = df_focus_peak[i,SNP_ID]
		peakID = df_focus_peak[i,Peak_ID]

		# get peak where this SNP has the lowest p-value
		ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",peakID, "_", SNP, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


# new file
file = "/sc/arion/projects/Microglia/Combined_RNAseq_ATACseq/GWAS_loci_investigations/GWAS_loci_analyses/Novel_Loci/AD_Loci_locations_RK_7_8_21.csv"
df_focus = fread(file)

# For each eQTL
df_focus_gene = df_focus[Gene_ID !='',]

for(i in 1:nrow(df_focus_gene)){

	SNP = df_focus_gene[i,SNP_ID]
	ensGene = df_focus_gene[i,Gene_ID]
	Symbol = df_focus_gene[i,GeneName]

	# get peak where this SNP has the lowest p-value
	peakID =	df_caQTL[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

	if( length(peakID) == 0){
		peakID = ''
	}
	file = paste0("figures/",Symbol, "_", SNP, "_fm.pdf")
	plot_triple(SNP, ensGene, peakID, window = 80000, file)
}


# For each caQTL
df_focus_peak = df_focus[Peak_ID !='',]

for(i in 1:nrow(df_focus_peak)){

		SNP = df_focus_peak[i,SNP_ID]
		peakID = df_focus_peak[i,Peak_ID]

		# get peak where this SNP has the lowest p-value
		ensGene = df_eqtl[Variant == SNP,.SD[which.max(log10.p.value),Gene]]

		file = paste0("figures/",peakID, "_", SNP, "_fm.pdf")
		plot_triple(SNP, ensGene, peakID, window = 80000, file)
}



```





